### PR TITLE
Added Typechecking to input tensor in RNN

### LINKED ISSUE #99998: Change misleading error message for type mismatch in LSTM forward
### 🚀 The feature, motivation and pitch

The following code:
```
import torch
lstm = torch.nn.LSTM(1, 200, 2, batch_first=True)
input = torch.zeros(20, 36, 1).long()
lstm(input, None)
```
Raises the error:
`RuntimeError: expected scalar type Long but found Float`
Which is misleading - because the solution is to pass input as float tensor
Suggested solution - perform a type check on the input in rnn.forward

If you think its worth something I'd be happy to implement it

### Alternatives

_No response_

### Additional context

_No response_

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @malfet

### PR DESCRIPTION
**Pull request by TamirCohen**
The input tensor of the RNN forward must be the same type as the weights.
While passing tensor of type long the error is:
    `RuntimeError: expected scalar type Long but found Float`
Which is misleading because it said to convert Something to Long, but the correct solution is to convert the input to Float (Which is the type of the weights).

The new error:
    `RuntimeError: input must have the type torch.float32, got type torch.int64`

Is correct and more verbose

Fixes #99998

### CODE REVIEW

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -209,6 +209,9 @@ def reset_parameters(self) -> None:
             init.uniform_(weight, -stdv, stdv)
 
     def check_input(self, input: Tensor, batch_sizes: Optional[Tensor]) -> None:
+        if input.dtype != self._flat_weights[0].dtype:
+            raise RuntimeError('input must have the type {}, got type {}'.format(
```
Comment (Skylion007) [inline torch/nn/modules/rnn.py:213]: Nit: this should be a ValueError, even if we don't consistently use that.

**Comment by TamirCohen** [comment]
@drisspg

**Review by drisspg — APPROVED** [review]
Looks good thanks!

**Pull request closed without merging**